modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java

```diff
@@ -944,15 +862,67 @@ public void finishCleanup() {
         }
     }
 
-    @Nullable VersionChain readVersionChain(RowId rowId) {
+    void throwExceptionIfStorageNotInRunnableState() {
+        StorageUtils.throwExceptionIfStorageNotInRunnableState(state.get(), this::createStorageInfo);
+    }
+
+    /**
+     * Searches version chain by row ID and converts the found version chain to the result if found.
+     *
+     * @param rowId Row ID.
+     * @param function Function for converting the version chain to a result, function is executed under the read lock of the page on which
+     *      the version chain is located. If the version chain is not found, then {@code null} will be passed to the function.
+     */
+    <T> @Nullable T findVersionChain(RowId rowId, Function<VersionChain, T> function) {
         try {
-            return versionChainTree.findOne(new VersionChainKey(rowId));
+            return versionChainTree.findOne(new VersionChainKey(rowId), new TreeRowClosure<>() {
+                @Override
+                public boolean apply(BplusTree<VersionChainKey, VersionChain> tree, BplusIo<VersionChainKey> io, long pageAddr, int idx) {
+                    return true;
+                }
+
+                @Override
+                public T map(VersionChain treeRow) {
+                    return function.apply(treeRow);
+                }
+            }, null);
         } catch (IgniteInternalCheckedException e) {
-            throw new StorageException("Error getting version chain: [rowId={}, {}]", e, rowId, createStorageInfo());
+            if (e.getCause() instanceof StorageException) {
+                throw (StorageException) e.getCause();
+            }
+
+            throw new StorageException("Row version lookup failed: [rowId={}, {}]", e, rowId, createStorageInfo());
         }
     }
 
-    void throwExceptionIfStorageNotInRunnableState() {
-        StorageUtils.throwExceptionIfStorageNotInRunnableState(state.get(), this::createStorageInfo);
+    /**
+     * Organizes external synchronization of update operations for the same version chain.
+     *
+     * <p>NOTE: When you try to execute in the closures on the pages of the tree, it leads to a deadlock.
+     */
+    protected <T> T inUpdateVersionChainLock(RowId rowId, Supplier<T> supplier) {
+        LockHolder<ReentrantLock> lockHolder = updateVersionChainLockByRowId.compute(rowId, (rowId1, reentrantLockLockHolder) -> {
```

Comment:
Maybe the entire map should be encapsulated into its own class, code would be really simple:
```
locker.lock(rowId);
...
locker.unlock(rowId);
```
Anyway, I see no distinction between readings and writings. Do we only lock stuff when we need updates? That's nice!

Reply:
At the moment, this map only locks updates.
I will try to describe it in the documentation for the class.

Reply:
Generally speaking, this would be cool to have as a separate class. But it's up to you. Having all these manipulations with map and counters is a boilerplate that can easily be hidden

Reply:
Perhaps for the moment I will leave everything as it is, because this map is used in one method.